### PR TITLE
[WIP] Make derive(Clone) memcpy on enum variants containing Copy types

### LINKED ISSUE #47796: derived Clone implementations for many-variant enums are unnecessarily large
The derived Clone implementations in Firefox weigh 192k [1], which is a lot.

Of that 192k, 79k of that is just for PropertyDeclaration::clone. PropertyDeclaration is a very large enum, but most of the variants are simple POD types. Ideally Rust would coalesce those together, and then generate special cases for the types that need it. Unfortunately, it appears that adding a single non-POD variant causes all the cases to be enumerated separately, as seen in the testcase at [2].

From IRC:

> eddyb	bholley: yeah I think this is a valid issue
> eddyb:	if you do generate that sort of "(partial) identity match" it's unlikely to have anything else collapse it
> eddyb:	LLVM might have a pass for switch-discriminant-dependent-values
> eddyb:	but it probably has serious limitations in practice
> eddyb:	bholley: but yeah this is one of those cases where not scattering in the first place is significantly easier than gathering later`

Ideally we'd do the same for PartialEq, since PropertyDeclaration::eq weighs another 61k.

[1] nm --print-size --size-sort --radix=d libxul.so | grep "\.\.Clone" | grep -v Cloned | cut -d " " -f 2 | awk '{s+=$1} END {print s}'
[2] https://play.rust-lang.org/?gist=0207af76d2e05acdbed913b7df96aa77&version=stable

### PR DESCRIPTION
**Pull request by Manishearth**
fixes https://github.com/rust-lang/rust/issues/47796

### CODE REVIEW

**Comment by Manishearth** [comment]
nowehere near ready for review, this was just me messing around. I might rewrite this to do something rather differently soon.

File: src/librustc/mir/mod.rs [hunk 1]
```diff
@@ -1520,6 +1520,9 @@ pub enum AggregateKind<'tcx> {
     /// active field number and is present only for union expressions
     /// -- e.g. for a union expression `SomeUnion { c: .. }`, the
     /// active field index would identity the field `c`
+    ///
+    /// For enums, the second field is the index of the variant
+    /// within AdtDef::fields
     Adt(&'tcx AdtDef, usize, &'tcx Substs<'tcx>, Option<usize>),
```
Comment (estebank) [inline src/librustc/mir/mod.rs]: I wonder if we should start moving to struct-like variants for anything that has more than 3 fields throughout the compiler...

**Review by estebank — COMMENTED** [review]
Will wait for further changes before properly reviewing, but we'll want somebody else besides me from the compiler team to look at it too.

**Comment by Manishearth** [comment]
Yeah I'm going to get review from eddyb later.

**Comment by Manishearth** [comment]
Superseded by #47867

**Pull request closed without merging**